File: examples/DrawDemo.c
// examples/DrawDemo.c
// Purpose: deterministic demo that prints a few u01 draws for selected engines.
// Matching R script: r-package/inst/examples/DrawDemo.R
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include "randompack.h"

static char *fmt1 = "%s seed=%d spawnkey=%d u01=%.17g %.17g\n";
static char *fmt2 = "%s counter=%d key=%d u01=%.17g %.17g\n";

static void draw_u01(const char *engine, int seed, int key) {
  randompack_rng *rng = randompack_create(engine);
  if (!rng) {
    printf("%s unavailable\n", engine);
    return;
  }
  randompack_seed(seed, (uint32_t*)&key, 1, rng);
  double x[2];
  randompack_u01(x, 2, rng);
  printf(fmt1, engine, seed, key, x[0], x[1]);
  randompack_free(rng);
}

static void draw_philox(int ctr0, int key0) {
  randompack_rng *rng = randompack_create("philox");
  if (!rng) {
    printf("philox unavailable\n");
    return;
  }
  uint64_t state[6] = {ctr0, 0, 0, 0, 0, 0};
  uint64_t key[2] = {key0, 0};
  randompack_set_state(state, 6, rng);
  randompack_philox_set_key(key, rng);
  double x[2];
  randompack_u01(x, 2, rng);
  printf(fmt2, "philox", (int)ctr0, (int)key0, x[0], x[1]);
  randompack_free(rng);
}

static void draw_squares(uint64_t ctr, int key) {
  randompack_rng *rng = randompack_create("squares");
  uint64_t state[2] = {ctr, 0};
  randompack_set_state(state, 2, rng);
  randompack_squares_set_key(key, rng);
  double x[2];
  randompack_u01(x, 2, rng);
  printf(fmt2, "squares", (int)ctr, key, x[0], x[1]);
  randompack_free(rng);
}

int main(void) {
  draw_u01("x256++", 42, 2);
  draw_u01("pcg64", 42, 2);
  draw_philox(1, 42);
  draw_u01("chacha20", 42, 2);
  draw_squares(1, 123456789);
  return 0;
}

File: r-package/inst/examples/DrawDemo.R
# inst/examples/DrawDemo.R
# Purpose: deterministic demo that prints a few unif draws for selected engines.
# Matching C program: examples/DrawDemo.c

suppressMessages(library(randompack))

fmt1 <- "%s seed=%d spawnkey=%d unif=%.17g %.17g\n"
fmt2 <- "%s counter=%d key=%d unif=%.17g %.17g\n"

u64_halves <- function(x) {
  if (!is.numeric(x) || length(x) != 1L || is.na(x))
    stop("x must be a length-1 numeric")
  x <- as.numeric(x)
  if (x < 0 || x > 18446744073709551615)
    stop("x must be in [0, 2^64-1]")
  lo <- x %% 4294967296
  hi <- floor(x / 4294967296)
  c(lo, hi)
}

draw_u01 <- function(engine, seed, key) {
  rng <- randompack_rng(engine)
  if (is.null(rng)) stop("engine creation failed")
  rng$seed(seed, as.integer(key))
  x <- rng$unif(2)
  cat(sprintf(fmt1, engine, seed, key, x[1], x[2]))
  invisible(NULL)
}

draw_philox <- function(counter0, key0) {
  rng <- randompack_rng("philox")
  if (is.null(rng)) stop("engine creation failed")
  counter <- c(u64_halves(counter0), 0, 0, 0, 0, 0, 0)
  key <- c(u64_halves(key0), 0, 0)
  rng$philox_set_state(counter, key)
  x <- rng$unif(2)
  cat(sprintf(fmt2, "philox", counter0, key0, x[1], x[2]))
  invisible(NULL)
}

draw_squares <- function(counter, key) {
  rng <- randompack_rng("squares")
  counter_vec <- u64_halves(counter)
  key_vec <- u64_halves(key)
  rng$squares_set_state(counter_vec, key_vec)
  x <- rng$unif(2)
  cat(sprintf(fmt2, "squares", counter, key, x[1], x[2]))
  invisible(NULL)
}

main <- function() {
  draw_u01("x256++", 42, 2)
  draw_u01("pcg64", 42, 2)
  draw_philox(1, 42)
  draw_u01("chacha20", 42, 2)
  draw_squares(1, 123456789)
  invisible(NULL)
}

main()
